plugins/net_plugin/net_plugin.cpp

```diff
@@ -4453,63 +4606,71 @@ namespace eosio {
             ++num_clients;
          } else {
             ++num_peers;
          }
          if (update_p2p_connection_metrics) {
-            fc::unique_lock g_conn((*it)->conn_mtx);
-            boost::asio::ip::address_v6::bytes_type addr = (*it)->remote_endpoint_ip_array;
+            fc::unique_lock g_conn(c->conn_mtx);
+            boost::asio::ip::address_v6::bytes_type addr = c->remote_endpoint_ip_array;
             g_conn.unlock();
             net_plugin::p2p_per_connection_metrics::connection_metric metrics{
-                 .connection_id = (*it)->connection_id
+                 .connection_id = c->connection_id
                , .address = addr
-               , .port = (*it)->get_remote_endpoint_port()
-               , .accepting_blocks = (*it)->is_blocks_connection()
-               , .last_received_block = (*it)->get_last_received_block_num()
-               , .first_available_block = (*it)->get_peer_start_block_num()
-               , .last_available_block = (*it)->get_peer_head_block_num()
-               , .unique_first_block_count = (*it)->get_unique_blocks_rcvd_count()
-               , .latency = (*it)->get_peer_ping_time_ns()
-               , .bytes_received = (*it)->get_bytes_received()
-               , .last_bytes_received = (*it)->get_last_bytes_received()
-               , .bytes_sent = (*it)->get_bytes_sent()
-               , .last_bytes_sent = (*it)->get_last_bytes_sent()
-               , .connection_start_time = (*it)->connection_start_time
-               , .log_p2p_address = (*it)->log_p2p_address
+               , .port = c->get_remote_endpoint_port()
+               , .accepting_blocks = c->is_blocks_connection()
+               , .last_received_block = c->get_last_received_block_num()
+               , .first_available_block = c->get_peer_start_block_num()
+               , .last_available_block = c->get_peer_head_block_num()
+               , .unique_first_block_count = c->get_unique_blocks_rcvd_count()
+               , .latency = c->get_peer_ping_time_ns()
+               , .bytes_received = c->get_bytes_received()
+               , .last_bytes_received = c->get_last_bytes_received()
+               , .bytes_sent = c->get_bytes_sent()
+               , .last_bytes_sent = c->get_last_bytes_sent()
+               , .block_sync_bytes_sent = c->get_block_sync_bytes_sent()
+               , .connection_start_time = c->connection_start_time
+               , .log_p2p_address = c->log_p2p_address
             };
             per_connection.peers.push_back(metrics);
          }
 
-         if (!(*it)->socket_is_open() && (*it)->state() != connection::connection_state::connecting) {
-            if (!(*it)->incoming()) {
-               if (!(*it)->resolve_and_connect()) {
-                  it = connections.erase(it);
+         if (!c->socket_is_open() && c->state() != connection::connection_state::connecting) {
+            if (!c->incoming()) {
+               g.unlock();
```

Comment:
Once the lock is unlocked the iterator can no longer be considered valid.

Reply:
Reworked to never hold an iterator without the lock.